File: javascript/src/repositories/billing/customer.ts
import type { QueryFn } from "../types.js";

const BILLING_CUSTOMER_GET_SQL =
  "SELECT user_id FROM bursar.billing_customers WHERE provider = $1 AND provider_customer_id = $2";

/** Repository for billing customer operations. */
export class BillingCustomerRepository {
  constructor(private query: QueryFn) {}

  /** Upsert a billing customer record. */
  async upsert(
    provider: string,
    providerCustomerId: string,
    userId: string,
    email: string | null,
  ): Promise<void> {
    await this.query(
      `INSERT INTO bursar.billing_customers (provider, provider_customer_id, user_id, email)
       VALUES ($1, $2, $3, $4)
       ON CONFLICT (provider, provider_customer_id) DO UPDATE SET
         user_id = EXCLUDED.user_id,
         email = COALESCE(EXCLUDED.email, billing_customers.email),
         updated_at = now()`,
      [provider, providerCustomerId, userId, email],
    );
  }

  /** Fetch a user_id by provider identifiers. Returns null when no customer found.
   *
   * Guards against undefined/null user_id values that would produce the string
   * "undefined" or "null" instead of null.
   */
  async get(provider: string, providerCustomerId: string): Promise<string | null> {
    const rows = await this.query(BILLING_CUSTOMER_GET_SQL, [provider, providerCustomerId]);
    if (rows.length === 0) return null;
    const row = rows[0] as Record<string, unknown>;
    const userId = row.user_id;
    if (userId == null) return null;
    return String(userId);
  }

  /** Reverse lookup: find a customer record by user_id. Returns null if not found. */
  async getByUserId(
    userId: string,
    provider?: string | null,
  ): Promise<{ provider: string; providerCustomerId: string } | null> {
    if (provider) {
      const rows = await this.query(
        `SELECT provider, provider_customer_id
         FROM bursar.billing_customers
         WHERE user_id = $1 AND provider = $2
         ORDER BY updated_at DESC LIMIT 1`,
        [userId, provider],
      );
      if (rows.length === 0) return null;
      const row = rows[0] as Record<string, unknown>;
      return {
        provider: String(row.provider),
        providerCustomerId: String(row.provider_customer_id),
      };
    }
    const rows = await this.query(
      `SELECT provider, provider_customer_id
       FROM bursar.billing_customers
       WHERE user_id = $1
       ORDER BY updated_at DESC LIMIT 1`,
      [userId],
    );
    if (rows.length === 0) return null;
    const row = rows[0] as Record<string, unknown>;
    return {
      provider: String(row.provider),
      providerCustomerId: String(row.provider_customer_id),
    };
  }
}

File: python/src/bursar/repositories/billing/customer.py
from __future__ import annotations

from typing import Any

from bursar.repositories._types import DbQuery
from bursar.repositories._utils import validate_non_empty


class BillingCustomerRepository:
    """Repository for billing customer operations."""

    def __init__(self, execute: DbQuery) -> None:
        self._execute = execute

    def upsert(
        self,
        provider: str,
        provider_customer_id: str,
        user_id: str,
        email: str | None,
    ) -> dict[str, Any]:
        """Insert or update a billing customer record via RPC."""
        validate_non_empty(provider, "provider")
        validate_non_empty(provider_customer_id, "provider_customer_id")
        validate_non_empty(user_id, "user_id")
        rows = self._execute(
            "SELECT public.upsert_billing_customer(%s, %s, %s::uuid, %s) AS result",
            [provider, provider_customer_id, user_id, email],
        )
        if not rows:
            return {"status": "ok"}
        row = rows[0]
        result = row.get("result") if isinstance(row, dict) else row[0]
        if isinstance(result, dict):
            return result
        return {"status": "ok"}

    def get(self, provider: str, provider_customer_id: str) -> str | None:
        validate_non_empty(provider, "provider")
        validate_non_empty(provider_customer_id, "provider_customer_id")
        rows = self._execute(
            "SELECT user_id FROM public.billing_customers WHERE provider = %s AND provider_customer_id = %s",
            [provider, provider_customer_id],
        )
        if not rows:
            return None
        row = rows[0]
        if isinstance(row, dict):
            user_id = row.get("user_id")
            return str(user_id) if user_id is not None else None
        return None

    def get_by_user_id(
        self,
        user_id: str,
        provider: str | None = None,
    ) -> dict[str, Any] | None:
        validate_non_empty(user_id, "user_id")
        if provider is not None:
            validate_non_empty(provider, "provider")
            rows = self._execute(
                """SELECT provider, provider_customer_id
                   FROM public.billing_customers
                   WHERE user_id = %s AND provider = %s
                   ORDER BY updated_at DESC LIMIT 1""",
                [user_id, provider],
            )
        else:
            rows = self._execute(
                """SELECT provider, provider_customer_id
                   FROM public.billing_customers
                   WHERE user_id = %s
                   ORDER BY updated_at DESC LIMIT 1""",
                [user_id],
            )
        if not rows:
            return None
        row = rows[0]
        return row if isinstance(row, dict) else None
